orchard_pallas.py

```diff
@@ -0,0 +1,231 @@
+            x = self.exp(tm1d2 + 1)
+            # b <- x \omega = a \omega^2
+            b = self.exp(T)
+            y = S
+
+            # 7: while b != 1 do
+            while b != self.ONE:
+                # 8: Find least integer k >= 0 such that b^(2^k) == 1
+                k = 1
+                b2k = b * b
+                while b2k != self.ONE:
+                    b2k = b2k * b2k
+                    k += 1
+                assert k < y
+
+                # 9:
+                # w <- z^(2^(y-k-1))
+                for _ in range(0, y - k - 1):
+                    c = c * c
+                # x <- xw
+                x = x * c
+                # z <- w^2
+                c = c * c
+                # b <- bz
+                b = b * c
+                # y <- k
+                y = k
+            assert x * x == self
+            return x
+        elif a == self.MINUS_ONE:
+            return None
+        return self.ZERO
+
+
+class Scalar(FieldElement):
+    def __init__(self, s, strict=False):
+        FieldElement.__init__(self, Scalar, s, q, strict=strict)
+
+    def __str__(self):
+        return 'Scalar(%s)' % self.s
+
+Fp.ZERO = Fp(0)
+Fp.ONE = Fp(1)
+Fp.MINUS_ONE = Fp(-1)
+
+assert Fp.ZERO + Fp.ZERO == Fp.ZERO
+assert Fp.ZERO + Fp.ONE == Fp.ONE
+assert Fp.ONE + Fp.ZERO == Fp.ONE
+assert Fp.ZERO - Fp.ONE == Fp.MINUS_ONE
+assert Fp.ZERO * Fp.ONE == Fp.ZERO
+assert Fp.ONE * Fp.ZERO == Fp.ZERO
+
+
+#
+# Point arithmetic
+#
+
+PALLAS_B = Fp(5)
+
+class Point(object):
+    @staticmethod
+    def rand(rand):
+        while True:
+            data = rand.b(32)
+            p = Point.from_bytes(data)
+            if p is not None:
+                return p
+
+    @staticmethod
+    def from_bytes(buf):
+        assert len(buf) == 32
+        if buf == bytes([0]*32):
+            return Point.identity()
+
+        y_sign = buf[31] >> 7
+        buf = buf[:31] + bytes([buf[31] & 0b01111111])
+        try:
+            x = Fp.from_bytes(buf)
+        except ValueError:
+            return None
+
+        x3 = x * x * x
+        y2 = x3 + PALLAS_B
+
+        y = y2.sqrt()
+        if y is None:
+            return None
+
+        if y.s % 2 != y_sign:
+            y = Fp.ZERO - y
+
+        return Point(x, y)
+
+    def __init__(self, x, y):
+        self.x = x
+        self.y = y
+        self.is_identity = False
+
+    def identity():
+        p = Point(Fp.ZERO, Fp.ZERO)
+        p.is_identity = True
+        return p
+
+    def __neg__(self):
+        if self.is_identity:
+            return self
+        else:
+            return Point(Fp(self.x.s), -Fp(self.y.s))
+
+    def __add__(self, a):
+        if self.is_identity:
+            return a
+        elif a.is_identity:
+            return self
+        else:
+            (x1, y1) = (self.x, self.y)
+            (x2, y2) = (a.x, a.y)
+
+            if x1 != x2:
+                # <https://core.ac.uk/download/pdf/10898289.pdf> section 4.1
+                λ = (y1 - y2) / (x1 - x2)
```

Comment:
What version of Python is required to parse these lambda chars?

Reply:
Actually this works on Python 2.7 with the `# -*- coding: utf8 -*-` line. But this file is explicitly declared as Python 3 anyway.